tests/blackbox.py

```diff
@@ -440,3 +477,425 @@ def dryrun_one(
+
+        # --- setup reporter and stats --- #
+
+        meth_name: Any
+
+        def msg4simulate() -> str:
+            return f"""user provide message={msg}
+call_strat={type(approval_strat)}
+{meth_name=}
+{oc=}
+{call_cfg=}
+{is_app_create=}
+{len(self.predicates[meth_name])=}
+{stats["method_combo_count"]=}
+{stats["dryrun_count"]=}
+{stats["assertions_count"]=}
+"""
+
+        def update_stats(meth, num_preds):
+            stats[str(meth)] += num_dryruns
+            stats["method_combo_count"] += 1
+            stats["dryrun_count"] += num_dryruns
+            stats["assertions_count"] += num_dryruns * num_preds
+
+        # ---- APPROVAL PROGRAM SIMULATION ---- #
+        approval_strat = None
+        call_cfg = None
+        approve_sim = None
+        if not omit_approval_call:
+            approval_strat = ABICallStrategy(
+                json.dumps(contract.dictify()),
+                approval_args_strat_type,
+                num_dryruns=num_dryruns,
+                abi_args_mod=approval_abi_args_mod,
+            )
+            double_check_at_least_one_method = False
+            for meth_name, meth_cfg in method_configs.items():
+                sig = approval_strat.method_signature(meth_name)
+                approve_sim = Simulation(
+                    self.algod,
+                    ExecutionMode.Application,
+                    approval_teal,
+                    self.predicates[meth_name],
+                    abi_method_signature=sig,
+                    identities_teal=model_approval_teal,
+                    validation=executor_validation,
+                )
+
+                for oc_str, call_cfg in asdict(meth_cfg).items():
+                    oc = as_on_complete(oc_str)
+
+                    def simulate_approval(on_create):
+                        tp: TxParams = deepcopy(txn_params)
+                        tp.update_fields(
+                            TxParams.for_app(is_app_create=on_create, on_complete=oc)
+                        )
+                        sim_results = approve_sim.run_and_assert(
+                            approval_strat, txn_params=tp, msg=msg4simulate()
+                        )
+                        assert sim_results.succeeded
+                        if meth_name not in self.results:
+                            self.results[meth_name] = {}
+                        self.results[meth_name][(on_create, oc)] = sim_results
+                        update_stats(meth_name, len(self.predicates[meth_name]))
+
+                    # weird walrus is_app_create := ... to fill closure of msg4simulate()
+                    if call_cfg & CallConfig.CALL:
+                        double_check_at_least_one_method = True
+                        simulate_approval(is_app_create := False)
+
+                    if call_cfg & CallConfig.CREATE:
+                        double_check_at_least_one_method = True
+                        simulate_approval(is_app_create := True)
```

Comment:
nit: I noticed a bit of handwaving between `is_app_create := True/False` here and `simulate_approval`. For the sake of readability, do you want to declare `is_app_create` ahead?

Reply:
see https://github.com/algorand/pyteal/pull/634/commits/deca0acba00c971c10dbe1e7c7a78a60cf7c641d

Reply:
thank you! looks good